File: frontend/src/types/project.ts
export type ProjectStatus =
  | "draft"
  | "script_ready"
  | "generating"
  | "preview_ready"
  | "exported"
  | "failed";

export type ProjectAssetType = "character_image" | "scene_image" | "tts";

export type GenerationJobStatus = "pending" | "running" | "failed" | "completed";

export type StoryCharacter = {
  id: string;
  name: string;
  role?: string | null;
  appearance?: string | null;
  notes?: string | null;
  voice_id?: string | null;
  voice_label?: string | null;
};

export type StoryScene = {
  id: string;
  name: string;
  description: string;
  visual_prompt?: string | null;
};

export type StoryBeat = {
  id: string;
  title: string;
  summary: string;
  emotion?: string | null;
};

export type StoryDialogue = {
  speaker: string;
  speaker_id?: string | null;
  content: string;
};

export type StoryShot = {
  id: string;
  title: string;
  scene_id?: string | null;
  beat_id?: string | null;
  summary: string;
  narration?: string | null;
  dialogues: StoryDialogue[];
  character_ids: string[];
  camera?: string | null;
  emotion?: string | null;
  expected_duration?: number | null;
};

export type StoryboardDraft = {
  version: number;
  characters: StoryCharacter[];
  scenes: StoryScene[];
  beats: StoryBeat[];
  shots: StoryShot[];
  suggested_duration?: number | null;
  notes: string[];
};

export type ProjectStoragePaths = {
  project_dir: string;
  project_file: string;
  storyboard_file: string;
  assets_dir: string;
  exports_dir: string;
};

export type ProjectAssetPayload = {
  provider?: string | null;
  model?: string | null;
  prompt?: string | null;
  remote_url?: string | null;
  script?: string | null;
  voice_style?: string | null;
  voice_id?: string | null;
  audio_format?: string | null;
  target_id?: string | null;
  asset_path?: string | null;
  image_local_path?: string | null;
  audio_local_path?: string | null;
  manual_override?: boolean;
  content?: string | null;
  label?: string | null;
  segments?: Record<string, unknown>[];
  target_kind?: string | null;
};

export type ProjectAssets = Partial<Record<ProjectAssetType, Record<string, ProjectAssetPayload>>>;

export type GenerationJob = {
  id: string;
  project_id: string;
  job_type: ProjectAssetType;
  target_id: string;
  provider_key: string;
  status: GenerationJobStatus;
  payload?: Record<string, unknown> | null;
  result?: Record<string, unknown> | null;
  error_message?: string | null;
  celery_task_id?: string | null;
  retry_count: number;
  created_at: string;
  updated_at: string;
};

export type PreviewTimelineAudioTrack = {
  track_type: "tts" | "bgm";
  segment_type: "narration" | "dialogue" | "gap";
  target_id: string;
  shot_id: string;
  source_path?: string | null;
  label: string;
  speaker?: string | null;
  voice_id?: string | null;
  start_offset_seconds: number;
  duration_seconds: number;
};

export type PreviewTimelineShot = {
  order: number;
  shot_id: string;
  title: string;
  duration_seconds: number;
  subtitle: string;
  narration?: string | null;
  scene_asset_path?: string | null;
  character_asset_paths: string[];
  audio_segments: PreviewTimelineAudioTrack[];
};

export type PreviewTimeline = {
  project_id: string;
  status: "ready";
  preview_file: string;
  total_duration: number;
  shot_count: number;
  render_mode: "placeholder" | "mixed" | "real_assets";
  scene_asset_count: number;
  audio_asset_count: number;
  updated_at: string;
  shots: PreviewTimelineShot[];
  audio_tracks: PreviewTimelineAudioTrack[];
};

export type ExportStatus = {
  project_id: string;
  status: "idle" | "running" | "completed" | "failed";
  preview_file?: string | null;
  export_file?: string | null;
  total_duration?: number | null;
  shot_count: number;
  render_mode: "placeholder" | "mixed" | "real_assets";
  scene_asset_count: number;
  audio_asset_count: number;
  error_message?: string | null;
  updated_at: string;
};

export type ProjectDetail = {
  id: string;
  title: string;
  source_text: string;
  genre?: string | null;
  style_template?: string | null;
  target_duration: number;
  voice_style?: string | null;
  status: ProjectStatus;
  created_at: string;
  updated_at: string;
  storage: ProjectStoragePaths;
  storyboard?: StoryboardDraft | null;
  assets: ProjectAssets;
};

export type ProjectListItem = {
  id: string;
  title: string;
  genre?: string | null;
  style_template?: string | null;
  target_duration: number;
  voice_style?: string | null;
  status: ProjectStatus;
  updated_at: string;
  created_at: string;
  storyboard_ready: boolean;
  shot_count: number;
  asset_count: number;
};

export type CreateProjectPayload = {
  title: string;
  source_text: string;
  genre?: string;
  style_template?: string;
  target_duration: number;
  voice_style?: string;
};

export type ReplaceProjectAssetPayload = {
  content: string;
  label?: string;
};

File: backend/app/schemas/project.py
from __future__ import annotations

from datetime import datetime
from typing import Literal, Optional

from pydantic import BaseModel, ConfigDict, Field

from app.models.project import ProjectStatus
from app.models.storyboard import StoryboardDraft


class CreateProjectRequest(BaseModel):
    title: str = Field(min_length=1, max_length=200)
    source_text: str = Field(min_length=1)
    genre: Optional[str] = Field(default=None, max_length=100)
    style_template: Optional[str] = Field(default=None, max_length=100)
    video_style: Optional[str] = Field(default=None, max_length=100)
    target_duration: int = Field(default=60, ge=15, le=300)
    voice_style: Optional[str] = Field(default=None, max_length=100)
    aspect_ratio: Optional[str] = Field(default=None, max_length=16)
    bgm_style: Optional[str] = Field(default=None, max_length=100)


class ProjectStoragePaths(BaseModel):
    project_dir: str
    project_file: str
    storyboard_file: str
    video_plan_file: str
    assets_dir: str
    exports_dir: str


class VideoSegmentVariantResponse(BaseModel):
    variant_id: str
    label: str
    status: Literal["pending", "submitted", "running", "completed", "failed"] = "pending"
    video_asset_path: Optional[str] = None
    remote_video_url: Optional[str] = None
    thumbnail_asset_path: Optional[str] = None
    provider_task_id: Optional[str] = None
    provider_file_id: Optional[str] = None
    raw_status: Optional[str] = None
    error_message: Optional[str] = None
    duration_seconds: Optional[float] = None


class VideoSegmentResponse(BaseModel):
    segment_id: str
    title: str
    summary: str
    prompt: str
    selected_variant_id: Optional[str] = None
    variants: list[VideoSegmentVariantResponse] = Field(default_factory=list)


class VideoGenerationPlanResponse(BaseModel):
    segment_count: int
    global_style_bible: dict[str, str] = Field(default_factory=dict)
    segments: list[VideoSegmentResponse] = Field(default_factory=list)


class ProjectAssetPayload(BaseModel):
    provider: Optional[str] = None
    model: Optional[str] = None
    prompt: Optional[str] = None
    remote_url: Optional[str] = None
    remote_video_url: Optional[str] = None
    script: Optional[str] = None
    voice_style: Optional[str] = None
    voice_id: Optional[str] = None
    audio_format: Optional[str] = None
    target_id: Optional[str] = None
    asset_path: Optional[str] = None
    video_local_path: Optional[str] = None
    image_local_path: Optional[str] = None
    audio_local_path: Optional[str] = None
    manual_override: bool = False
    content: Optional[str] = None
    label: Optional[str] = None
    segments: list[dict[str, object]] = Field(default_factory=list)
    target_kind: Optional[str] = None


class TimelineAudioTrackResponse(BaseModel):
    track_type: Literal["tts", "bgm"]
    segment_type: Literal["narration", "dialogue", "gap"]
    target_id: str
    shot_id: str
    source_path: Optional[str] = None
    label: str
    speaker: Optional[str] = None
    voice_id: Optional[str] = None
    start_offset_seconds: float = 0
    duration_seconds: float = 0


class ReplaceProjectAssetRequest(BaseModel):
    content: str = Field(min_length=1, max_length=1000)
    label: Optional[str] = Field(default=None, max_length=100)


class SelectSegmentVariantRequest(BaseModel):
    variant_id: str = Field(min_length=1)


class TimelineShotResponse(BaseModel):
    order: int
    shot_id: str
    title: str
    duration_seconds: int
    subtitle: str
    narration: Optional[str] = None
    video_asset_path: Optional[str] = None
    scene_asset_path: Optional[str] = None
    character_asset_paths: list[str] = Field(default_factory=list)
    audio_segments: list[TimelineAudioTrackResponse] = Field(default_factory=list)


class PreviewTimelineResponse(BaseModel):
    project_id: str
    status: Literal["ready"]
    preview_file: str
    total_duration: int
    shot_count: int
    render_mode: Literal["placeholder", "mixed", "real_assets", "video_segments"]
    scene_asset_count: int = 0
    audio_asset_count: int = 0
    updated_at: datetime
    shots: list[TimelineShotResponse] = Field(default_factory=list)
    audio_tracks: list[TimelineAudioTrackResponse] = Field(default_factory=list)


class ExportStatusResponse(BaseModel):
    project_id: str
    status: Literal["idle", "running", "completed", "failed"]
    preview_file: Optional[str] = None
    export_file: Optional[str] = None
    total_duration: Optional[int] = None
    shot_count: int = 0
    render_mode: Literal["placeholder", "mixed", "real_assets", "video_segments"] = "placeholder"
    scene_asset_count: int = 0
    audio_asset_count: int = 0
    error_message: Optional[str] = None
    updated_at: datetime


class UpdateScriptDraftRequest(BaseModel):
    storyboard: StoryboardDraft


class ProjectDetailResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: str
    title: str
    source_text: str
    genre: Optional[str]
    style_template: Optional[str]
    video_style: Optional[str] = None
    target_duration: int
    voice_style: Optional[str]
    aspect_ratio: Optional[str] = None
    bgm_style: Optional[str] = None
    status: ProjectStatus
    created_at: datetime
    updated_at: datetime
    storage: ProjectStoragePaths
    video_plan: Optional[VideoGenerationPlanResponse] = None
    storyboard: Optional[StoryboardDraft] = None
    assets: dict[str, dict[str, ProjectAssetPayload]] = Field(default_factory=dict)


class ProjectListItemResponse(BaseModel):
    id: str
    title: str
    genre: Optional[str] = None
    style_template: Optional[str] = None
    video_style: Optional[str] = None
    target_duration: int
    voice_style: Optional[str] = None
    aspect_ratio: Optional[str] = None
    bgm_style: Optional[str] = None
    status: ProjectStatus
    updated_at: datetime
    created_at: datetime
    storyboard_ready: bool = False
    shot_count: int = 0
    asset_count: int = 0
